zng/docs/spec.md

```diff
@@ -350,121 +470,126 @@ previously defined type, the appropriate typedef control code must
 be re-emitted
 (and note that the typedef may now be assigned a different ID).
 
-### 3.1.3 Compressed Value Message Block
 
-Following a header byte of 0x86 is a compressed value message block.
-Such a block comprises a compressed sequence of value messages.  The
-sequence must not include control messages.
-
-A compressed value message block is encoded as follows:
-```
-------------------------------------------------------------------------------
-|0x86|<format><uncompressed-length>|<compressed-length>|<compressed-messages>|
-------------------------------------------------------------------------------
 
-```
-where
-* `<format>`, a `uvarint`, identifies the algorthim used to compress the
-  message sequence
-* `<uncompressed-length>`, a `uvarint`, is the length in bytes of the
-  uncompressed message sequence
-* `<compressed-length>`, a `uvarint`, is the length in bytes of `<compressed-messages>`
-* `<compressed-messages>` is the compressed value message sequence
+### 3.2 Value Messages
 
-Values for `<format>` are defined in the
-[ZNG compression format specification](./compression-spec.md).
+Following a header byte in the range `0x00-0xde` is a ZNG value.
+The header byte indicates the type ID of the value.  If the type ID
+is larger than `0xde`, then the type ID is "escaped" with the value `0xdf`
+and the actual type ID is encoded as a `uvarint` of the difference
+of the type ID less the constant `0xdf`.
+
+It is an error for a value to reference a type ID that has not been
+previously defined by a typedef scoped to the stream in which the value
+appears.
+
+The value is encoded in the subsequent bytes using a "tag-encoding" scheme
+that captures the structure of both primitive types and the recursive
+nature of complex types.  This structure is encoded
+explicitly in every value and the boundaries of each value and its
+recursive nesting can be parsed without knowledge of the type or types of
+the underlying values.  This admits an efficient implementation
+for traversing the values, inclusive of recursive traversal of complex values,
+whereby the inner loop need not consult and interpret the type ID of each element.
 
-### 3.2 Value Messages
+#### 3.2.1 Tag-Encoding of Values
 
-Following a header byte with bit 7 zero is a `typed value`
-with a `uvarint7` encoding its length.
+Each value is prefixed with a "tag" that defines:
+* whether it is a primitive or complex value,
+* whether it is a nil value, and
```

Comment:
I think some discussion of what `nil` means would be helpful. Likewise for `null`.